src/com/amazon/ionelement/api/Element.kt

```diff
@@ -0,0 +1,257 @@
+    val values: Collection<IonElement>
+}
+
+/**
+ * Represents an Ion list or s-expression.
+ *
+ * Includes common functional list operations such as [head] and [tail].
+ */
+interface SeqElement : ContainerElement {
+    override val values: List<IonElement>
+
+    /**
+     * Returns the string representation of the symbol in the first element of this container.
+     *
+     * If the first element is not a symbol or this container has no elements, throws [IonElectrolyteException],
+     */
+    val tag get() = this.head.symbolValue
+
+    /**
+     * Returns the first element of this container.
+     *
+     * If this container has no elements, throws [IonElectrolyteException].
+     */
+    val head: IonElement
+        get() =
+            when (this.values.size) {
+                0 -> ionError(this, "Cannot get head of empty container")
+                else -> this.values.first()
+            }
+
+    /**
+     * Returns a sub-list containing all elements of this container except the first.
+     *
+     * If this container has no elements, throws [IonElectrolyteException].
+     */
+    val tail: List<IonElement> get()  =
+        when (this.values.size) {
+            0 -> ionError(this, "Cannot get tail of empty container")
+            else -> this.values.subList(1, this.values.size)
+        }
+}
+
+/**
+ * Represents an Ion list.
+ *
+ * Includes no additional functionality over [SeqElement], but serves to provide additional type safety when
+ * working with elements that must be Ion lists.
+ */
+interface ListElement : SeqElement
+
+/**
+ * Represents an Ion s-expression.
+ *
+ * Includes no additional functionality over [SeqElement], but serves to provide additional type safety when
+ * working with elements that must be Ion s-expressions.
+ */
+interface SexpElement : SeqElement
+
+/**
+ * Represents an Ion struct.
+ *
+ * Includes functions for accessing the fields of a struct.
+ */
+interface StructElement : ContainerElement {
+
+    /** This struct's unordered collection of fields. */
+    val fields: Collection<IonStructField>
+
+    /** A list of the unique field names contained within this struct. */
+    val fieldNames: List<String>
+
+    /**
+     * Retrieves the value of the first field found with the specified name.
+     * The search order is undefined.
+     *
+     * @throws IonElectrolyteException If there are no fields with the specified [fieldName].
+     */
+    operator fun get(fieldName: String): IonElement
+
+    /**
+     * Retrieves the first field found with the specified name or `null` if no such field exists.
+     *
+     * This is the same as [get], except it returns `null` instead of throwing if the field doesn't exist.
+     *
+     * QUESTION/TODO:  I shied a way from the "OrNull" suffix for this one since we use that term for different purpose
+     * in IonElement.  Open to suggestions for other names if this one isn't ideal.
+     */
+    fun findOne(fieldName: String): IonElement?
```

Comment:
There's an interesting question here since there are three scenarios w need to handle...

- The field exists and is a null value
- The field exists and is a null value
- The field does not exist

I think the next revision of this should have

- `get` throw if the field doesn't exist or is an Ion null
- `getOrNull` throw only if the field doesn't exist

Reply:
On deeper thought about his...

I think if `fieldNames` above is a `Set<String>` then we really don't need this function.